tools/autograd/templates/python_variable_methods.cpp

```diff
@@ -528,29 +529,56 @@ static PyObject * THPVariable_storage_type(PyObject* self, PyObject* arg)
   END_HANDLE_TH_ERRORS
 }
 
+// Cases:
+// 1) arg matches self -> no-op
+// 2) arg is true, self doesn't requires_grad -> do the equivalent of detach() and set_requires_grad.
+// 3) arg is false, self requires_grad -> valid if self is a leaf.
+static Variable as_requires_grad(Tensor self, bool requires_grad) {
+  auto& self_ = static_cast<torch::autograd::Variable&>(self);
+  auto self_requires_grad = self_.requires_grad();
+  if (self_requires_grad == requires_grad) {
+    return self_;
+  } else if (requires_grad) {
+    auto res = make_variable(self_.data(), /*requires_grad=*/true);
+    res.set_version_counter(self_.version_counter());
+    return res;
+  } else {
+    if (!self_.is_leaf()) {
+      throw std::runtime_error(requires_grad_leaf_error(requires_grad));
+    }
+    self_.set_requires_grad(requires_grad);
+    return self_;
+  }
+}
+
 static PyObject * THPVariable_to(PyObject* self, PyObject* args, PyObject* kwargs)
 {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
-    "to(Device device, ScalarType dtype=None)",
-    "to(ScalarType dtype)",
-    "to(Tensor other)",
+    "to(*, bool requires_grad)",
+    "to(Device device, ScalarType dtype=None, *, bool requires_grad=None)",
+    "to(ScalarType dtype, *, bool requires_grad=None)",
+    "to(Tensor other)",  // lacks requires_grad because the default is ambiguous between self and other
```

Comment:
I can't see how not adding `requires_grad` here makes it unambiguous. The only thing you removed is the ability to specify it explicitly, but this actually never would be ambiguous anyway. I think we should just default to whatever the `requires_grad` of `self` is (which is what this implementation does I think), and use the given value otherwise. WDYT?

Reply:
Having a `self.to(other, requires_grad=???)` makes the behavior a different from the other functions we have that have a requires_grad keyword:
- `tensor.new_zeros(...)` (returns a tensor with requires_grad=False by default) and 
- `torch.zeros_like(tensor)` (returns a tensor with requires_grad=False by default). 

Which is okay but makes me feel a uneasy

Reply:
Maybe the right way to think about it is x.to(*params, requires_grad=r) is equivalent to x.to(*params).to(requires_grad=r).